pkg/controller/volume/attachdetach/reconciler/reconciler.go

```diff
@@ -219,11 +223,17 @@ func (rc *reconciler) reconcile() {
 				continue
 			}
 
-			// Trigger detach volume which requires verifing safe to detach step
+			// insert a detach operation start time into cache if needed for the volume for metrics reporting
+			// note that the plugin name has been set to "" as there is no visibility at this moment to which
+			// plugin/csi driver would be conducting the operation. It will be populated later when detach
+			// operation function is created and plugin name then could be decided
+			rc.operationStartTimeCache.AddIfNotExist(createOperationUniqueKey(attachedVolume.VolumeName, attachedVolume.NodeName), "", "volume_detach")
```

Comment:
It's hard to understand/maintain when the start of the metric exists here (in the reconciler) and I believe the "end" of the same metric is after the success in the operation generator. Could we also cache the start time in the start of the function in the operation generator and just have it not overwrite the start time if it is on the n'th run-through? 

I'm also not exactly sure what the goal is so maybe i'm misunderstanding :)

/cc @msau42

Reply:
The goal is to record end to end attach/detach operation. However, if a detach operation needs to happen before an attach can happen, i.e., a volume is currently attached to a different node when the controller tries to attach it to a specific node, the detach operation latency will NOT be included the attach operation latency.
You are not misunderstanding, and the end of the operation would be when the final success happens in operation generator. It is also possible to have the cache set the first time an attach/detach operation has been generated in operation generator.
Putting it in reconcile though, makes it easier to manage the lifecycle of the cached item, i.e., when an attach/detach operation needs to be aborted while trying. Also this is also more extendable in the future if we'd like to report "waiting_time" metrics, i.e., the total time has spend on waiting on a volume to be detached before attach can happen.

The cached start timestamp will eventually be removed when Operation generator's attach function returns success, and thats when the metric will be reported.

Reply:
@davidz627 I have moved the insertion into operation generator as the time_waiting piece and error reporting still remains to be discussed. Please take a look. thnx